include/seastar/core/gate.hh

```diff
@@ -47,21 +51,36 @@ public:
 class gate {
     size_t _count = 0;
     std::optional<promise<>> _stopped;
+
+#ifdef SEASTAR_GATE_HOLDER_DEBUG
+    size_t _holders = 0;
+
+    void assert_not_held() const noexcept {
+        assert(!_holders && "gate moved with outstanding holders");
+    }
+#else   // SEASTAR_GATE_HOLDER_DEBUG
+    void assert_not_held() const noexcept {}
+#endif  // SEASTAR_GATE_HOLDER_DEBUG
+
 public:
     gate() = default;
     gate(const gate&) = delete;
     gate(gate&& x) noexcept
-        : _count(std::exchange(x._count, 0)), _stopped(std::exchange(x._stopped, std::nullopt)) {}
+        : _count(std::exchange(x._count, 0)), _stopped(std::exchange(x._stopped, std::nullopt)) {
+        x.assert_not_held();
+    }
     gate& operator=(gate&& x) noexcept {
         if (this != &x) {
             assert(!_count && "gate reassigned with outstanding requests");
+            x.assert_not_held();
             _count = std::exchange(x._count, 0);
             _stopped = std::exchange(x._stopped, std::nullopt);
         }
         return *this;
     }
     ~gate() {
         assert(!_count && "gate destroyed with outstanding requests");
+        assert_not_held();
```

Comment:
nit: this would print "gate moved ..." while technically the gate is not moved here

Reply:
Fixed in [dc5afad](https://github.com/scylladb/seastar/commit/dc5afadbd5deec5c069029016fd8c399551b45f7)